test/cases/testdata/refheads/test-refs-as-rule-heads.yaml

```diff
@@ -0,0 +1,74 @@
+cases:
+- modules:
+  - |
+    package test
+
+    a.b.c.p { true }
+  note: 'refheads/complete: direct query'
+  query: data.test.a.b.c.p = x
+  want_result:
+  - x: true
+- modules:
+  - |
+    package test
+
+    a.b.c.p { true }
+  note: 'refheads/complete: full package extent'
+  query: data.test = x
+  want_result:
+  - x:
+      a:
+        b:
+          c:
+            p: true
+- modules:
+  - |
+    package test
+
+    a.b.c.p = 1
+    q = 0
+    a.b.d = 3
+
+    p {
+      q == 0
+      data.test.a.b.c.p == 1
```

Comment:
given that `q` can be accessed directly, without the `data.test.` prefix, within the same module; it's easy to assume that `a.b.c.p` could be too. But this is not the case, right?

Reply:
> But this is not the case, right?

Exactly. I think this is in line with how we deal with multiple modules: when there's a mod with `package a.b.c` and one with in `package a`, the latter's rego can't use `b.c.d` to reference rule `d` in `package a.b.c` either. But we could discuss this decision!

Reply:
IMO, it makes sense to be able to directly reference something in the same module/file as where it was declared. I can't say for certain, but I think a lot of people would try to ref `a.b.c.p` before they tried `data.test.a.b.c.p`.

Someone could of course declare an object in that file `a.b := {"c": {"p": 42}}`, which would make `q := a.b.c.p` ambiguous .. But I don't think the `data.<module package>` prefix of calling `q := data.test.a.b.c.p` instead actually helps with that, right?

Reply:
I'll double-check tomorrow, but some cases are captured by the conflict check.

Reply:
You would be able to do that if you did

```rego
package test
import data.test.a

a.b.c.p = 1
p {
  a.b.c.p == 1
}
```
...but this never occurred to me before. I'll have to see if it works or just breaks. 🔍

Reply:
Works as-is:

```
$ opa eval -fpretty -d x.rego data               
{
  "test": {
    "a": {
      "b": {
        "c": {
          "p": 1
        }
      }
    },
    "p": true
  }
}```

Reply:
My expectation was that `a` would be visible inside of `package test` just like any other rule/variable. I realize that the implementation means it doesn't work this way but I think we should change that.

```
package test

# these should behave the same
x := {"foo": {"bar": {"baz": 7}}}
y["foo"] := {"bar": {"baz": 7}}
z.foo.bar.baz := 7

p {
  x.foo.bar.baz == 7
  y.foo.bar.baz == 7
  z.foo.bar.baz == 7  # currently you can't say this, but IMO, that's confusing
}
```

I'd rather not require users add an import for this to work. I'll take a look at the implementation.

Reply:
My expectation was that `a` would be visible inside of `package test` just like any other rule/variable. I realize that the implementation means it doesn't work this way but I think we should change that.

```
package test

# these should behave the same
x := {"foo": {"bar": {"baz": 7}}}
y["foo"] := {"bar": {"baz": 7}}
z.foo.bar.baz := 7

p {
  x.foo.bar.baz == 7
  y.foo.bar.baz == 7
  z.foo.bar.baz == 7  # currently you can't say this, but IMO, that's confusing
}
```

I'd rather not require users add an import for this to work. I'll take a look at the implementation.

Reply:
👍 makes sense. Also the same is true for x.foo.bar and x.foo and x (y, z respectively), where the scope of resolving z.foo or z.foo.bar would be broader, as it might include other files' and modules' contributions to the tree.